File: py_payment_api/app/services/forwarder/abstract/multiforwarder.py
from typing import Dict, TypeVar, Generic
from .forwarder import Forwarder
from enum import StrEnum

T = TypeVar('T', bound=StrEnum)

class MultiForwarder(Generic[T], Forwarder):
    def __init__(self, forwarders: Dict[T, Forwarder]):
        self.forwarders = forwarders

    def _on_forward_event(self, event_data: dict) -> None:
        api_type = event_data.get('api_type')
        if not api_type or api_type not in self.forwarders:
            raise ValueError("Invalid or missing API type")
        
        forwarder = self.forwarders[api_type]
        forwarder.forward_event(event_data)

File: tsPaymentApi/src/services/forwarder/abstract/MultiForwarder.ts
import { Forwarder } from './Forwarder';

export abstract class MultiForwarder<T extends string> implements Forwarder {
    protected forwarders: Map<T, Forwarder>;

    constructor(forwarders: Map<T, Forwarder>) {
        this.forwarders = forwarders;
    }

    public async forwardEvent(eventData: Record<string, any>): Promise<void> {
        const apiType = this.getApiType(eventData);
        const forwarder = this.forwarders.get(apiType);
        
        if (!forwarder) {
            throw new Error(`No forwarder found for API type: ${apiType}`);
        }

        await forwarder.forwardEvent(eventData);
    }

    protected abstract getApiType(eventData: Record<string, any>): T;
} 